Add Python 3.11 to supported: test and build whl for it

yarikoptic opened the pull request:
Originally issue came up to be untable to test heudiconv against 3.11 since installation from sources in
https://github.com/nipy/heudiconv/actions/runs/4126346722/jobs/7255036166 has failed.  That is how I discovered that the .whl was not built for pylibjpeg-libjpeg . Then I saw that 3.11 is not yet tested against etc.

I guess it would need a release to get that whl done/available to resolve our heudiconv issue.

mrbean-bremen commented on .github/workflows/release-wheels.yml:
```diff
@@ -17,7 +17,7 @@ jobs:
       name: Install Python
       with:
-        python-version: '3.10'
+        python-version: '3.10', '3.11'
```
Will this work? Will it actually install 2 versions?

yarikoptic replied:
sorry for that -- my `sed` went too wild, will fix now.

yarikoptic commented:
3.11 build fails with
```
  libjpeg/_libjpeg.cpp:196:12: fatal error: longintrepr.h: No such file or directory
```

cython had fix for that in https://github.com/cython/cython/pull/4428 for 0.29.25 milestone. Current release 0.29.33 . The file here produced with 

```
❯ head -n 1 `git grep -l longintrepr.h`
/* Generated by Cython 0.29.20 */
```

although newer version is required already

```
❯ grep cython pyproject.toml
    "setuptools == 60.5", "cython == 0.29.26", "wheel == 0.37",
```
looking at `git log` for that file, seems to be no automated/formalized way to regenerate it... I will manually re-cythonize now using the cython I have (0.29.32-2+b1 on debian)

mrbean-bremen commented:
I think you also have to add the 3.11 wheels to the `build-wheels` matrix `release-wheels.yaml` - currently it is only used in the tests. You could test that part by temporary disabling the upload step and run it on "pull-request" instead of "release".

yarikoptic commented:
eh - seems to be missing numpy whl for OSX for 3.11 ? odd
```
      Collecting numpy==1.22
        Downloading numpy-1.22.0.zip (11.3 MB)
...
      Building wheels for collected packages: numpy
        Building wheel for numpy (pyproject.toml): started
        Building wheel for numpy (pyproject.toml): still running...
        Building wheel for numpy (pyproject.toml): still running...
        Building wheel for numpy (pyproject.toml): still running...
        Building wheel for numpy (pyproject.toml): finished with status 'error'
        error: subprocess-exited-with-error
```

odd since https://pypi.org/project/numpy/#files lists 

[numpy-1.24.2-cp311-cp311-macosx_11_0_arm64.whl ](https://files.pythonhosted.org/packages/01/04/a8b0bb5ffd6b36cb9ff9b67ca6966d55c4a9fdb40ace81a2b33d1559c3b7/numpy-1.24.2-cp311-cp311-macosx_11_0_arm64.whl)(13.8 MB [view hashes](https://pypi.org/project/numpy/#copy-hash-modal-8cdc123c-a3e8-4ba3-ad6d-c4cd772016fc))
Uploaded Feb 5, 2023 cp311

[numpy-1.24.2-cp311-cp311-macosx_10_9_x86_64.whl ](https://files.pythonhosted.org/packages/1b/a7/9582b169194a05642fcd05026b2e55fa7539230bfc28de7e13f116b0cd0b/numpy-1.24.2-cp311-cp311-macosx_10_9_x86_64.whl)(19.8 MB [view hashes](https://pypi.org/project/numpy/#copy-hash-modal-897371f9-a064-41e2-931b-123b24987400))
Uploaded Feb 5, 2023 cp311

ah - different numpy version, for https://pypi.org/project/numpy/1.22.0/#files indeed there is no cp311... let's boost to what version? make it python version dependent?

mrbean-bremen commented:
> ah - different numpy version, for https://pypi.org/project/numpy/1.22.0/#files indeed there is no cp311... let's boost to what version? make it python version dependent?

Actually the NumPy release notes say that Python 3.7 is only supported until 1.21, so it may indeed make sense to make it version-dependent.

yarikoptic commented on pyproject.toml:
```diff
@@ -2,5 +2,6 @@
     "setuptools == 60.5", "cython == 0.29.26", "wheel == 0.37",
     "numpy==1.20; python_version=='3.7'",
-    "numpy==1.22; python_version>='3.8'",
+    "numpy==1.22; python_version>='3.8',<'3.11'",
```
@jwodder -- is this the correct specification for a range of python versions here?

yarikoptic replied:
pip gave me the answer - No ;)
```
× Can not process file:///home/yoh/proj/misc/pylibjpeg-libjpeg
╰─> This package has an invalid `build-system.requires` key in pyproject.toml.
    It contains an invalid requirement: "numpy==1.22; python_version>='3.8',<'3.11'"

note: This is an issue with the package mentioned above, not pip.
hint: See PEP 518 for the detailed specification.
```

yarikoptic replied:
nevermind, figured it out, fix is coming

yarikoptic commented:
![image](https://user-images.githubusercontent.com/39889/219209543-e836993a-d1d0-48ea-ab0a-7289e42163c4.png)

mrbean-bremen commented on .github/workflows/release-wheels.yml:
```diff
@@ -17,18 +17,13 @@ jobs:
-      uses: actions/upload-artifact@v3
-      with:
-        name: wheels
-        path: ./dist
```
This part is needed if you want the tests to run on the packages.

mrbean-bremen commented:
Same problem with cython now in the package tests...

yarikoptic commented:
I am not sure I am fully following the logic of that test there, but I see that it seems downloading released (1.3.3) version of it - so it would not be surprising that it would fail since not re-cythoned:

```
2023-02-16T07:14:59.8026483Z Looking in links: dist/
2023-02-16T07:14:59.9454082Z Collecting pylibjpeg-libjpeg
2023-02-16T07:15:00.1103046Z   Downloading pylibjpeg-libjpeg-1.3.3.tar.gz (682 kB)
2023-02-16T07:15:00.1739492Z      ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 682.9/682.9 kB 12.1 MB/s eta 0:00:00
2023-02-16T07:15:00.2852314Z   Installing build dependencies: started
2023-02-16T07:16:07.5826306Z   Installing build dependencies: still running...
2023-02-16T07:17:13.3128152Z   Installing build dependencies: still running...
2023-02-16T07:18:14.1362236Z   Installing build dependencies: still running...
2023-02-16T07:19:12.5097445Z   Installing build dependencies: finished with status 'done'
2023-02-16T07:19:12.5109224Z   Getting requirements to build wheel: started
2023-02-16T07:19:13.0302386Z   Getting requirements to build wheel: finished with status 'done'
2023-02-16T07:19:13.0345497Z   Preparing metadata (pyproject.toml): started
2023-02-16T07:19:13.4227798Z   Preparing metadata (pyproject.toml): finished with status 'done'
2023-02-16T07:19:13.4303064Z Requirement already satisfied: numpy>=1.22 in ./testwhl/lib/python3.11/site-packages (from pylibjpeg-libjpeg) (1.24.2)
2023-02-16T07:19:13.4323456Z Building wheels for collected packages: pylibjpeg-libjpeg
2023-02-16T07:19:13.4330945Z   Building wheel for pylibjpeg-libjpeg (pyproject.toml): started
```

mrbean-bremen commented:
> I am not sure I am fully following the logic of that test there, but I see that it seems downloading released (1.3.3) version of it

It takes the version that has been created before and saved as artifact (I have re-added this step, because without it no tests can run). The number is taken from the version of the package,  which has not been increased since last release (you could do this, if you want, to avoid confusion).

yarikoptic commented:
I am pretty positive that it is working with prior/non-cythoned version of that file as if it downloaded a release tarball, look at error line number:

```
  libjpeg/_libjpeg.cpp:196:12: fatal error: longintrepr.h: No such file or directory
        196 |   #include "longintrepr.h"
            |            ^~~~~~~~~~~~~~~
      compilation terminated.
      error: command '/usr/bin/gcc' failed with exit code 1
      [end of output]
```
and what we have in master (196 too -- match)

```
❯ git co master
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
❯ git grep -n longintrepr
libjpeg/_libjpeg.cpp:196:  #include "longintrepr.h"
```

and in this branch - 262.

```
❯ git co -
Switched to branch 'enh-3.11'
Your branch is up to date with 'gh-yarikoptic/enh-3.11'.
❯ git grep -n longintrepr
libjpeg/_libjpeg.cpp:262:    #include "longintrepr.h"
```
and I don't see in that log any indication that the file might have been re-cythonized with older cython:

```
2023-02-16T07:19:14.1082975Z       running build_ext
2023-02-16T07:19:14.1083346Z       skipping 'libjpeg/_libjpeg.cpp' Cython extension (up-to-date)
```

but hey -- may be it was re-cythonized in some prior step?  way too many runs... need to `grep`... will create myself a https://github.com/con/tinuous copy of the logs first ....

mrbean-bremen commented:
Hm - did you change anything, apart from the version? The tests are now passing...

yarikoptic commented:
nope, didn't change anything besides version. Staring at the logs now

yarikoptic commented:
don't know why but FWIW indeed with version which does not correspond the released one the log does NOT say it is downloading that tarball:

```
 Looking in links: dist/
-Collecting pylibjpeg-libjpeg
-  Downloading pylibjpeg-libjpeg-1.3.3.tar.gz (682 kB)
-     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 682.9/682.9 kB 12.1 MB/s eta 0:00:00
+Processing ./dist/pylibjpeg-libjpeg-1.3.999.tar.gz
   Installing build dependencies: started
-  Installing build dependencies: still running...
-  Installing build dependencies: still running...
-  Installing build dependencies: still running...
   Installing build dependencies: finished with status 'done'
   Getting requirements to build wheel: started
   Getting requirements to build wheel: finished with status 'done'
```

yarikoptic commented:
I think overall it might be boiling down to the `pip` resolving ambiguity of having the same version locally and from PyPI in favor of PyPI while running that `python -m pip install -U --find-links dist/ pylibjpeg-libjpeg` .  

Note that also the name is "Install from package wheels and test" but it installs from the source, not from the wheel. ... why that test doesn't do smth like

```
python -m pip install dist/*.whl
```
to ensure installing from the built whl?

yarikoptic commented:
Let's try that -- pushed.  FWIW, I can promise to not be evil and not to do anything evil, and try to be supercarefull here if you invite me to the project so workflows could just run without you blessing them. but that is ok if you prefer not too -- I'm not looking for more work ;-) thought this would be easy PR ;-)

mrbean-bremen commented:
Thanks for all your work! As soon as your changes are merged you will be able to run your workflows yourself, this is only a problem for the first PR. Anyway, sleep time over here, but it looks as if this is practically done now...

yarikoptic commented:
> As soon as your changes are merged you will be able to run your workflows yourself, this is only a problem for the first PR.

hm, well, not sure since it would mean that another *insane me* who isn't  part of the project could fix tiny something in the workflow on first PR and then in subsequent PR just make it rule the world without interference!  but anyways -- hopefully not applicable since after this PR the life will just be trouble-less and we never talk again.

yarikoptic commented:
unfortunately 
```
WARNING: Requirement 'dist/pylibjpeg-libjpeg-*.whl' looks like a filename, but the file does not exist
```
and `download-artifacts` doesn't say what it downloaded and I don't have access to the artifacts I believe, so can't check. and was lazy to add `ls` to see what we have there, will do now so you could trigger the build first time in the morning. I will also think positive and kick away that commit with version boost.

mrbean-bremen commented:
> hm, well, not sure since it would mean that another insane me who isn't part of the project could fix tiny something in the workflow on first PR and then in subsequent PR just make it rule the world without interference!

Well, I guess we have to live with that risk... 😁 

Anyway, as I wrote - your help is very much appreciated, and I'm looking forward to more contributions, if you are so inclined. 
I'm not the project owner here, and I didn't write a single line of the code, so I want to leave decisions about collaborators to @scaramallion , who will hopefully be back here soon. As I wrote, you can can always contribute, and after the first PR, the restriction of not being able to run the workflow will be gone (this had been introduced by GitHub some time ago, and is indeed a bit inconvenient given the different timezones people are living in).

BTW, [pylibjpeg-openjpeg](https://github.com/pydicom/pylibjpeg-openjpeg)  also lacks 3.11 support, in case you are getting bored 😈 

> and download-artifacts doesn't say what it downloaded and I don't have access to the artifacts I believe, so can't check

You actually [have access to the artifacts](https://docs.github.com/en/actions/managing-workflow-runs/downloading-workflow-artifacts), the only problem is that you need to have different names for the artifacts, otherwise will see only the latest one (in this, one of the Python 3.11 wheels, as all artifacts are called `wheel`).

mrbean-bremen commented:
Ah, and now you have removed storing the artifacts again, so you won't get any anyway...

jmsmkn commented:
Sorry, I completely missed this PR! I went the other way around and added support for Python 3.11 in `pylibjpeg-openjpeg` (https://github.com/pydicom/pylibjpeg-openjpeg/pull/58), then faced the same problem with this library so applied similar changes in #44 as once I found the root cause the solution was straightforward, I'll leave that open as an alternative.

mrbean-bremen commented:
Thanks @jmsmkn - in this case you are probably the right person to review this one!

yarikoptic commented:
> Ah, and now you have removed storing the artifacts again, so you won't get any anyway...

maaaan, don't do that to me.  Just after I think I regained some sanity chasing some issue with git-annex/our HPC filesystem now this, where
- I see your commit 8e9cc78a8b126d2f0a2d0d391ca330003d37c852 in the history
- I see 
```shell
 ❯ git grep -p 'Store art'
.github/workflows/release-wheels.yml=jobs:
.github/workflows/release-wheels.yml:    - name: Store artifacts
```
- but I do see that "Store" in the diff as being removed
```shell
❯ git diff master...150aede -- .github/workflows/release-wheels.yml | grep Store
-    - name: Store artifacts
```
!!  but apparently the reason is that you had it twice!

```shell
❯ git co -
Switched to branch 'master'
Your branch is up to date with 'origin/master'.
❯ git grep -p 'Store art'
.github/workflows/release-wheels.yml=jobs:
.github/workflows/release-wheels.yml:    - name: Store artifacts
.github/workflows/release-wheels.yml:    - name: Store artifacts
```
so indeed one was removed, but never added, i.e. neither I nor you have never re-added it !  that would explain so many things now in retrospect... I have readded that too, and rebased/squashed all the remove/readd/readd commits for a bit saner log history.

mrbean-bremen commented:
😆 
Well, looks like it is coming together now 🤞. You could also add the 2 missing changes from the other PR (`requirements.txt` and updated cython version). Funny that just after I mentioned that the same has to be done for `libjpeg-openjpeg` somebody has already doing it... Well, another pair of eyes certainly can help.

mrbean-bremen commented on .github/workflows/release-wheels.yml:
```diff
@@ -206,7 +225,8 @@ jobs:
         python -m pip install git+https://github.com/pydicom/pylibjpeg-data
-        python -m pip install -U --find-links dist/ pylibjpeg-libjpeg
+        ls -l dist
+        python -m pip install -U dist/*.whl
```
I think this will not work as it will try to install all wheels, while with the old command it will look in that directory and find the matching one.

yarikoptic replied:
I thought that it would be just 1 whl, but indeed we would have one per each platform all in there, uff.
I didn't like previous one since as we discovered, it is fine to find none and  just install from online while downloading that thing for matching version.  We do not want to temporarily boost version for testing here  since it might hide the issue later IMHO.  Ideally we should have got only that particular whl here but the day is short so I indeed will just kick this commit out.

mrbean-bremen replied:
I think there is an option in pip to prevent it from looking online, I forgot what it's called (not at a computer right now).

yarikoptic replied:
might be as generic as setup `http_proxy=http://127.0.0.1:9/ https_proxy=http://127.0.0.1:9/` env vars, but might be too brute force . may be @asmacdo or @jwodder know a more gentle way from top of their heads?

jwodder replied:
The `--no-index` option to `pip install` will cause it to ignore the package index and only use `--find-links`.

mrbean-bremen replied:
Exactly, that is what I meant, thanks.

yarikoptic commented:
it was fun indeed but probably we should have just left it to @jmsmkn  to do it cleanly in one shot ;)

mrbean-bremen commented:
> it was fun indeed but probably we should have just left it to @jmsmkn to do it cleanly in one shot ;)

Ah no - you got this far and certainly learned a few things here, and who said that fun doesn't matter 😁

jmsmkn commented:
> Thanks @jmsmkn - in this case you are probably the right person to review this one! EDIT: Oh, I may not have understood correctly - which PR do you want to leave open?
> 
> The PRs look very similar, except that in this one `requiements.txt` has not been adapted, and the cython version in `pyproject.toml` has not been adapted. So I would propose to finish this PR (with the temporary changes to check the created packages left in until it works), and you may review this one, while I'll check the other one in `libjpeg-openjpeg` that you did.

There is one more change in #45 to the `install_requires` in `setup.py` to the numpy dependencies per python version, unsure if that is necessary as that may be covered by the numpy itself, but made that change for consistency with the rest.

> > it was fun indeed but probably we should have just left it to @jmsmkn to do it cleanly in one shot ;)
> Ah no - you got this far and certainly learned a few things here, and who said that fun doesn't matter 😁

Indeed, it's the journey and not the destination, sorry for stepping on any toes! Like I said, completely missed this one after making the changes for openjpeg.

mrbean-bremen commented:
@jmsmkn - thank you, and no, you were certainly not stepping on any toes!

@yarikoptic - if you have verified the packages to work, you could add the publishing part back and adapt the NumPy version as mentioned by @jmsmkn, and I will happily merge it.

jmsmkn commented:
@yarikoptic you will also need to bump the version of `cibuildwheel`, see https://github.com/pydicom/pylibjpeg-openjpeg/pull/59.

yarikoptic commented:
done (removed the TEMP commit, adjusted numpy versioning)

yarikoptic commented:
BTW, if you follow up with a release for all the goodness to enable 3.11 builds of dependents (`pydicom` -> `heudiconv`), would be much appreciated.
